### PR TITLE
Add TSPLIB95-importer

### PR DESCRIPTION
**Pull request by HannesWell**
I created an importer to read symmetric `travelling salesman problem` instances from files in the `TSPLIB95` format.
Files in TSPLIB95 format contain exactly defined TSP instances. As stated in the Javadoc of the importer there are large TSPLIB95 libraries containing problem instances of different magnitudes provided by the University of [Heidelberg](http://comopt.ifi.uni-heidelberg.de/software/TSPLIB95/) or [Waterloo](http://www.math.uwaterloo.ca/tsp/data/index.html).

The instances provided there can be used as unified benchmarks or tests for TSP algorithms. 
Additionally the importer could be used to improve the tests of JGraphT's TSP algorithms. At the moment some tests use randomly created instances that are therefore different for each test run. Such instances could be stored in a TSPLIB95 file and then load each time the test is executed to have the same instance every time. Also TSP test instances with special properties could be defined to test certain aspects or corner cases of an algorithm.

----

- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/Become-a-Contributor>
- [x] I read and understood <https://github.com/jgrapht/jgrapht/wiki/How-to-make-your-first-%28code%29-contribution>
- [x] I added [unit tests](https://github.com/jgrapht/jgrapht/wiki/Unit-testing)
- [x] I added [documentation](https://github.com/jgrapht/jgrapht/wiki/How-to-write-documentation)
- [x] I followed the [Coding and Style Conventions](https://github.com/jgrapht/jgrapht/wiki/Coding-and-Style-Conventions)
- [x] I **have not** modified `HISTORY.md` or `CONTRIBUTORS.md`
- [x] I ensured that [the git commit message is a good one](https://github.com/joelparkerhenderson/git_commit_message)

### CODE REVIEW

File: jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java [hunk 1]
```diff
@@ -0,0 +1,735 @@
...
+ * cities of different countries.
+ * </p>
+ * 
+ * @author Hannes Wellmann
+ * @param <V> the type of vertices created by this importer
+ *
+ */
+public class TSPLIBImporter<V>
+    implements
+    GraphImporter<V, DefaultWeightedEdge>
```
Comment (d-michail) [inline jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java]: Unless absolutely necessary, we try to not fix the edge class for the user. Change this to a generic parameter `E`. This also affects your parallel streams loading later on in your code.

File: jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java [hunk 2]
```diff
@@ -0,0 +1,735 @@
...
+{
+    /**
+     * Container for data of an imported <em>TSPLIB95</em> file.
+     * 
+     * @author Hannes Wellmann
+     * @param <V> the type of vertices in the imported graph
+     */
+    public static class TSPLIBFileData<V>
```
Comment (d-michail) [inline jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java]: I would prefer a nicer name here. You are already inside class `TSPLIBImporter` which means that you do not have to prefix it with `TSPLIB` anymore. Additionally, since this information is data about data, I would use something like  a `Metadata` or something in this line.

File: jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java [hunk 3]
```diff
@@ -0,0 +1,735 @@
...
+        private String name;
+        private String type;
+        private String comment;
+        private Integer dimension;
+        private String edgeWeightType;
+        private String nodeCoordType;
+        private boolean hasDistinctLocations;
+        private Boolean hasDistinctEdges;
+
+        private Graph<V, DefaultWeightedEdge> graph;
```
Comment (d-michail) [inline jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java]: If you are only using this for `hasDistinctEdgesPerVertex` then ok, otherwise this is not needed.
The user provides the graph from the `importGraph` method, no need to include it back.

File: jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java [hunk 4]
```diff
@@ -0,0 +1,735 @@
...
+            return nodeCoordType;
+        }
+
+        /**
+         * Returns the complete {@link Graph} of location connections built from the imported file
+         * or null if no graph was imported.
+         * 
+         * @return the complete {@code Graph} of location connections
+         */
+        public Graph<V, DefaultWeightedEdge> getGraph()
```
Comment (d-michail) [inline jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java]: Remove this method.  We have a clear interface on all importers, which is to the `importGraph` method. No need to return the graph using two different ways.

File: jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java [hunk 5]
```diff
@@ -0,0 +1,735 @@
...
+            return graph;
+        }
+
+        /**
+         * Returns the {@link List} of {@link Integer Integers} describing vertex order of the tour
+         * from the imported file or null if no tour was imported.
+         * 
+         * @return the tour described by the list of vertex indices
+         */
+        public List<Integer> getTour()
```
Comment (d-michail) [inline jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java]: What does it mean to return a list of integers `List<Integer>`?  Vertices in JGraphT may be any type of object, you need to translate the integers to vertices.

File: jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java [hunk 6]
```diff
@@ -0,0 +1,735 @@
...
+        return elements;
+    }
+
+    // read of coordinates and graph computation
+
+    /**
+     * Read the complete {@link Graph} of node locations and returns true if all locations are
+     * distinct.
+     */
+    private boolean readCompleteNodeGraphData(
```
Comment (d-michail) [inline jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java]: Please merge methods `readCompleteNodeGraphData`, `readNodeCoordinateData` and `buildCompleteGraph` into a single method called 'readGraph` or something like this.

I know that best practices are to keep methods small, but in this case it is over complicating things.

File: jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java [hunk 7]
```diff
@@ -0,0 +1,735 @@
...
+    {
+        locations.forEach(graph::addVertex);
+
+        new CompleteGraphGenerator<V, E>().generateGraph(graph, null);
+
+        // Compute edge weights:
+        // Enable parallel streams. Synchronization not necessary, because DefaultWeightedEdge are
+        // IntrusiveEdges and modified directly.
+
+        graph.edgeSet().parallelStream().forEach(e -> {
```
Comment (d-michail) [inline jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java]: Remove the parallel stream here, in order to also avoid any problems with using a generic edge type `E`. 

In the last few years, we explicitly try to avoid code which makes assumptions about the user edge.

Comment (d-michail) [reply jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java]: Moreover, these are probably small instances since the TSP is a hard problem, giving no real motivation for parallel processing.

File: jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java [hunk 8]
```diff
@@ -0,0 +1,735 @@
...
+         * vector}.
+         * 
+         * @param vector from which an element value is to return
+         * @param index of the element to return
+         * @return the value of the specified element in the given vector
+         */
+        double getElement(V vector, int index);
+    }
+
+    private final Function<double[], V> vectorFactory;
```
Comment (d-michail) [inline jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java]: Completely remove the `vectorFactory` and the `elementGetter` abstractions. 

Graphs in JGraphT are build using the corresponding `vertexSupplier` inside the graph instance.

After you create the vertices using method `addVertex()`, you need to internally maintain a map from the vertices to the coordinates (read from file) of the vertex. Afterwards you need to iterate over all pair of vertices, acquire from the map the coordinates pair and call the corresponding distance method.

File: jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java [hunk 9]
```diff
@@ -0,0 +1,735 @@
...
+
+    /**
+     * Create a {@link TSPLIBImporter} that uses the given {@code vectorFactory} to create its
+     * vertices with certain element values and later uses the specified {@code elementGetter} to
+     * obtain the element values and compute the distances between the vertices.
+     * 
+     * @param vectorFactory used to create vector vertices with certain element values
+     * @param elementGetter used to obtain element values from the created vertex-objects
+     */
+    public TSPLIBImporter(Function<double[], V> vectorFactory, ElementGetter<V> elementGetter)
```
Comment (d-michail) [inline jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java]: No need to any parameter here.

File: jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java [hunk 10]
```diff
@@ -0,0 +1,735 @@
...
+    {
+        this.vectorFactory = vectorFactory;
+        this.elementGetter = elementGetter;
+    }
+
+    // import data
+
+    private TSPLIBFileData<V> lastImportData = null;
```
Comment (d-michail) [inline jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java]: Remove the `last` and the `=null` since Java does this automatically.

File: jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java [hunk 11]
```diff
@@ -0,0 +1,735 @@
...
+
+    /**
+     * Returns the {@link TSPLIBFileData} of the latest imported file or null, if no import
+     * completed yet or the latest import failed.
+     * 
+     * @return {@code TSPLIBFileData} of the latest import
+     */
+    public TSPLIBFileData<V> getLastImportData()
```
Comment (d-michail) [inline jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java]: Call this like `getMetadata()` or something similar.

File: jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java [hunk 12]
```diff
@@ -0,0 +1,735 @@
...
+     * {@link TSPLIBFileData} of the previous import can be obtained from
+     * {@link #getLastImportData()}.
+     * </p>
+     * 
+     * @param graph into which this importer writes, must be simple, weighted and empty or may be
+     *        null.
+     * @throws IllegalArgumentException if the specified {@code graph} is not simple or is not empty
+     */
+    @Override
+    public void importGraph(Graph<V, DefaultWeightedEdge> graph, Reader in)
```
Comment (d-michail) [inline jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java]: DefaultWeightedEdge -> E

File: jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java [hunk 13]
```diff
@@ -0,0 +1,735 @@
...
+    {
+        lastImportData = null;
+        vectorLength = -1;
+        if (graph == null) {
+            @SuppressWarnings("unchecked") Supplier<DefaultWeightedEdge> edgeSupplier =
```
Comment (d-michail) [inline jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java]: The graph is not supposed to be null here. It is not your responsibility to create a graph here.

Now I understand why you return the graph inside the metadata object. Please remove this pattern. We want to keep all importers working in a similar way.

File: jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java [hunk 14]
```diff
@@ -0,0 +1,735 @@
...
+        throws IOException
+    {
+        List<Integer> tour = dimension != null ? new ArrayList<>(dimension) : new ArrayList<>();
+
+        for (String line; (line = reader.readLine()) != null;) {
+            String lineContent = line.trim();
+            if ("-1".equals(lineContent)) {
+                break;
+            }
+            tour.add(Integer.valueOf(lineContent));
```
Comment (d-michail) [inline jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java]: You need to consult some `Map<Integer, V>` here which you probably need to build earlier in your code in order to create a tour which is a `List<V>`.

File: jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java (hunk 8 above)
Comment (HannesWell) [reply jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java]: All right. I agree with your remarks and will apply them in the next days.

The intention for the factory was to be able to set the coordinates read from the file to the vertices. I assumed that this importer would likely be used with some kind of vector data structure. But it looks like I was too specific here and I understand your point.

Nevertheless I would like to provide a user the possibility to have the read coordinates in the created vertices. What do you think about a `BiConsumer` to set the read coordinates into a vertex that a user can optionally supply in the constructor? The vertex would be created by the regular vertex supplier and the read coordinates would be stored Map as you suggested. Afterwards (if the setter is present) for each vertex its coordinates could be set to it with the setter. The subsequent edge computations would fully operate on the internal coordinate data structure of the importer.

Comment (HannesWell) [reply jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java]: Or alternatively I provide the unmodifiable Vertex-Coordinates-Map in the MetaData-object. I think is better because simpler.

File: jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java (hunk 5 above)
Comment (HannesWell) [reply jgrapht-io/src/main/java/org/jgrapht/nio/tsplib/TSPLIBImporter.java]: The problem is that a TSPLIB-file containing a tour does not need to contain the node coordinates defining the TSP-problem instance.
In the cases I know there is a file containing the `NODE_COORD_SECTION` with the coordinates of all nodes (where each node also has a number) and a separate file that contains for example the optimal tour or the best known tour. That tour is described by a List of node-numbers that correspond to the numbers in the file with the `NODE_COORD_SECTION`. It is possible but definitely not always the case that the node coordinates and the tour is contained in one file.

I agree that a List of vertices is better. I thought about multiple options to achieve this. The one I prefer is the following:

- Let `getTour()` return a `List<V>` that is only not null, if a file contains a `TOUR_SECTION`.
- create a separate method `importTour(List<V>)` that has a List of vertices as their parameter that is ordered according to the node numbers and returns the List of vertices in the order of the tour. This enables import from a file that only contains a `TOUR_SECTION` . I think this method should create a MetaData object too (but maybe this has to be discussed).
- To enable the latter method I would also add a `getOrderedVertices()` method to the MetaData class that returns the vertices imported from a file with `NODE_COORD_SECTION` in their "number"-order.

**Comment by HannesWell** [comment]
I revised the structure of the importer and adjusted the tests accordingly.
I agree with you, the importer was very specific and I tried to make it more generic and addressed the other points you mentioned. I think the new structure is suitable to fit the TSPLIB-format into a `GraphImporter` in a generic way.

**Comment by HannesWell** [comment]
Another question: Have you ever considered to include Hamcrest as dependency for tests? Their matchers are quiet handy in combination with JUnit. Or do you want to keep the number of dependencies small?

**Comment by jsichi** [comment]
Test dependencies are less of a concern, so I wouldn't be opposed to adding hamcrest matchers.

**Comment by d-michail** [comment]
I will take a look in the next few days. Sorry for the delay.

**Comment by HannesWell** [comment]
No problem. I'm also quite busy those days.

**Comment by d-michail** [comment]
Great! The final code ended up looking pretty nice!

**Comment by HannesWell** [comment]
Thank you. I'm happy to hear that.

**Comment by HannesWell** [comment]
Regarding the Hamcrest Matchers:
Do you want the Maven dependency to Hamcrest to be add in a separate PR or as a "side-effect" in a PR where they are used? In the first case the question would be whether it should be add to all jGraphT artefacts (and probably would not be used in the first place)?

**Comment by d-michail** [comment]
I guess better in a PR where they are also used.
